### PR TITLE
Detect log rotation and release renamed files

### PR DESCRIPTION
**Pull request by kenhys**
In the previous vesions, there was no way to detect log rotation
event when log file was rotated.

It causes that rotated log file was grabbed by DaemonLogger.
This was unexpected behavior.

This commit fixes by reopening log file correctly.

NOTE: This PR depends on inotify so only linux was supported.

### CODE REVIEW

**Comment by kenhys** [comment]
A bit more work is needed to fix https://github.com/fluent/fluentd/issues/2426#issuecomment-717646842 completely.

File: lib/serverengine/daemon_logger.rb [hunk 1]
```diff
@@ -58,6 +64,16 @@ def logdev=(logdev)
...
+      if ServerEngine::linux?
+        if logdev.is_a?(String)
+          @inotify.close if @inotify
+          @inotify = INotify::Notifier.new
+          @inotify.watch(logdev, :move_self) do |event|
+            close
+            reopen
+          end
+        end
+      end
```
Comment (cosmo0920) [inline lib/serverengine/daemon_logger.rb:76]: Can we extract as a method here?
Current implementation is a bit of messy for me.

File: serverengine.gemspec [hunk 2]
```diff
@@ -34,4 +34,7 @@ Gem::Specification.new do |gem|
     # windows dependencies
     gem.add_runtime_dependency("windows-pr", ["~> 1.2.5"])
   end
+  if /linux/ =~ fake_platform || (/linux/ =~ RUBY_PLATFORM && fake_platform.empty?)
```
Comment (cosmo0920) [inline serverengine.gemspec:37]: Should we add `x86-linux` and `x86_64-linux` as GEM_BUILD_FAKE_PLATFORM like as follows?
https://github.com/treasure-data/serverengine/blob/master/Rakefile#L19

File: lib/serverengine/daemon_logger.rb [hunk 3]
```diff
@@ -58,6 +64,16 @@ def logdev=(logdev)
         old_file_dev.close if old_file_dev
         @logdev = @file_dev
       end
+      if ServerEngine::linux?
+        if logdev.is_a?(String)
```
Comment (cosmo0920) [inline lib/serverengine/daemon_logger.rb:68]: Problematic `@file_dev` is always String class instance?
If not so, we should handle more class instance here.

**Comment by repeatedly** [comment]
serverengine has own log rotation mechanizm and it doesn't have this problem, right?
If this problem happens with only external log rotation tool, how about `rb-inofity` is optional?

**Comment by kenhys** [comment]
I've changed dependency to optional.

**Comment by repeatedly** [comment]
Looks good to me with optional approach :)

**Comment by kenhys** [comment]
I've rebased PR to follow recent master.

**Comment by ashie** [comment]
I've added CI for with and w/o rb-notify.
All checks are passed.

File: lib/serverengine/daemon_logger.rb [hunk 4]
```diff
@@ -76,9 +85,51 @@ def add(severity, message = nil, progname = nil, &block)
...
+        target = logdev
+      else
+        # ignore StringIO for some test cases
+        return
+      end
+      if target
+        @inotify.close if @inotify
+        @inotify = INotify::Notifier.new
+        @inotify.watch(target, :move_self) do |event|
+          if logdev.respond_to?(:filename)
```
Comment (ashie) [inline lib/serverengine/daemon_logger.rb:119]: Should it be `@logdev` instead of `logdev`? @kenhys

Comment (kenhys) [reply lib/serverengine/daemon_logger.rb:119]: Thanks, it should be `@logdev` in this context.

**Comment by ashie** [comment]
> I've added CI for with and w/o rb-notify.
> All checks are passed.

No, still rb-inotify isn't recognized:

```
Pending:
  ServerEngine::DaemonLogger reopen log when path is renamed
    # not supported except Linux with inotify
    # ./spec/daemon_logger_spec.rb:182
  ServerEngine::DaemonLogger reopen log when path is renamed by external process
    # not supported except Linux with inotify
    # ./spec/daemon_logger_spec.rb:192
  ServerEngine::DaemonLogger reopen logger when file is renamed
    # not supported except Linux with inotify
    # ./spec/daemon_logger_spec.rb:205
  ServerEngine::MultiThreadServer raises SystemExit when all workers exit with specified code by unrecoverable_exit_codes
    # unrecoverable_exit_codes supported only for multi process workers
    # ./spec/multi_process_server_spec.rb:63
  ServerEngine::MultiThreadServer raises SystemExit immediately when a worker exits if 
```

File: lib/serverengine/daemon_logger.rb (hunk 4 above)
Comment (ashie) [reply lib/serverengine/daemon_logger.rb]: Fixed.

**Comment by ashie** [comment]
Hmm, when I test it on my local environment, sometime a test stalls.

It seems that it reproduced also on GitHub Actions:
https://github.com/treasure-data/serverengine/pull/106/checks?check_run_id=2652587049

@kenhys Could you take look this?

**Comment by kenhys** [comment]
In my local environment, `IO logger` test case seems to stall.
(Not identified the reason yet)

```
% bundle exec rspec -f documentation 2>&1 | tee ../take12.log

ServerEngine::BlockingFlag
  set and reset
  set! and reset! return whether it toggled the state
  wait_for_set timeout
  wait_for_reset timeout
  wait

ServerEngine::DaemonLogger
  reopen
  reset path
  default level is debug
  level set by int
  level set by string
  unknown level
  rotation
  IO logger
```

**Comment by kenhys** [comment]
It may be caused by `log.reopen!`.

```
it 'IO logger' do
    io = StringIO.new
    io.should_receive(:write)
    io.should_not_receive(:reopen)

    log = DaemonLogger.new(io)
    log.debug "stdout logging test"
    log.reopen! # <= STALLED HERE
  end
```

**Comment by kenhys** [comment]
I've misunderstand rspec workflow.

executing do block, call after (if exists), then show it ... label.
So, 'inter-process locking on rotation' test case is stalling. (next test case for 'IO logger')

**Comment by kenhys** [comment]
It seems that sometimes `Process.waitpid pid2` blocks.

https://github.com/treasure-data/serverengine/blob/master/spec/daemon_logger_spec.rb#L168

**Comment by kenhys** [comment]
As far as I know, only 'inter-process locking on rotation'  test case stalls. (frequency 2-3 stalls/10times)

**Comment by kenhys** [comment]
> As far as I know, only 'inter-process locking on rotation' test case stalls. (frequency 2-3 stalls/10times)

It tend to stall when  process(pid2) writes the limit (~30 bytes = log_rotate_age * log_rotate_size), proces(pid2) stucks.

**Comment by kenhys** [comment]
To clarify the issue:

* master (not stall)
* PR with rb-inotify (sometimes stalls with 'inter-process locking on rotation' test case)
* PR without rb-inotify (not stall)

**Comment by kenhys** [comment]
It tend to:

pass:
* `@inotify.watch` event is fired for process A and Process B (each process close @logdev and reopen it)

stall:
* `@inotify.watch` event is fired for process A only

**Comment by ashie** [comment]
I confirmed that it's blocked by `select` in `INotify::Notifier#readpratial`:

https://github.com/guard/rb-inotify/blob/010f1406f93d97c2fab06308786e4548c17fd588/lib/rb-inotify/notifier.rb#L318

```ruby
    def readpartial(size)
      readable, = select([@handle, @pipe.first])
```

I think the following usage described in [rb-inotify's README](https://github.com/guard/rb-inotify/blob/master/README.md) isn't usable in actual, because `select` is called twice against a event:
```ruby
 # Wait 10 seconds for an event then give up
 if IO.select([notifier.to_io], [], [], 10)
   notifier.process
 end
```

**Comment by ashie** [comment]
I've added a work around, but I'm not sure it's correct fix.
Further verification is needed.

**Comment by ashie** [comment]
I open another pull request to resolve this issue: #127

**Pull request closed without merging**